Volume observer pr4

nilsdeppe opened the pull request:
## Proposed changes

Adds volume observers that currently require a system-specific action. Once we have time (December-ish) we will need to go back and use the events and triggers infrastructure to properly handle observers in a more system-agnostic method.

Commit order is:
1. Add convenience wrapper around std::equal
2. Add support for calling threaded actions remotely
3. Add support for threaded_actions to mocking framework
4. Rename InvokeSimpleActionBase to InvokeActionBase
5. Factor out code from Test_RegisterElements
6. Fix name of execute_next_phase in observer component
7. Add ObserverWriter component
8. Add actions for observing volume data
9. Add trait observers::has_register_with_observer
10. Add support for registering with observers to DgElementArray
11. Add action for observing scalar wave solution and error.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Fix name of execute_next_phase in observer component

nilsdeppe committed:
Add trait observers::has_register_with_observer

This trait will be used to check if the Phases enum has a member
`RegisterWithObserver`

nilsdeppe committed:
Add support for registering with observers to DgElementArray

Currently we only register volume observers since this is all we have support
for.

nilsdeppe committed:
Add action for observing scalar wave solution and error.

nilsdeppe committed:
Add convenience wrapper around std::equal

nilsdeppe committed:
Factor out code from Test_RegisterElements

- The factored out code will be common to other observer tests
- Adds a missing include to Actions.hpp

nilsdeppe committed:
Add actions for observing volume data

wthrowe commented on tests/Unit/IO/Observers/ObserverHelpers.hpp:
```diff
@@ -0,0 +1,69 @@
+#include "tests/Unit/ActionTesting.hpp"
+
+// clang-tidy: we want to have the classes be only present in each object file,
+// not globally linked to.
```
This is an odd thing to do, so I would like the comment to explain why.

wthrowe commented on tests/Unit/IO/Observers/ObserverHelpers.hpp:
```diff
@@ -0,0 +1,69 @@
+  using chare_type = ActionTesting::MockArrayChare;
+  using array_index = size_t;
+  using const_global_cache_tag_list =
+      tmpl::list<observers::OptionTags::VolumeFileName>;
```
Acausal reference: this struct doesn't exist until the next commit.

wthrowe commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -0,0 +1,163 @@
+                    const observers::ArrayComponentId& array_component_id,
+                    std::vector<TensorComponent>&& in_received_tensor_data,
+                    const Index<Dim>& received_extents) noexcept {
+    db::mutate<Tags::VolumeDataLock, Tags::TensorData>(
```
Race condition: can't call `db::mutate` in parallel.

This is going to be a problem in general for threaded actions, since it means they can't fetch a lock object from the DataBox.  Even if we change to storing a pointer so we can access it through `db::get` we still can't modify the DataBox because there's a race between `db::mutate` and `db::get`.

nilsdeppe replied:
What if we prevent `db::mutate` and `db::get` from returning lock types, and add a separate function for retrieving lock types to the DataBox? This would allow you to get the lock, acquire it, mutate, release. The `get_lock` function I think then could get away with ignoring the `mutate_locked_box_` flag since the lock needs to be used separately anyway. I'll need to check how strongly typed Charm++'s lock is (It may be another of these pointer-to-void things), then we could wrap it in our own class to make the type checking possible.

nilsdeppe replied:
Never mind, there is a lock passed in to the threaded actions. I'm just forgetting how I wrote the code...

nilsdeppe replied:
Thinking about this a fair bit more (and having gotten the locks sorted out), I think what we actually want is to first collect on a group to avoid having to deal with locks, then collect the results from the groups on a nodegroup, followed by writing/further reduction/whatever we find is best. This approach will require acquiring a lock `num_elements_per_node/num_cores_per_node` fewer times, which will be between 10-1000.

wthrowe commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -0,0 +1,163 @@
+          const gsl::not_null<std::unordered_map<
+              observers::ObservationId,
+              std::unordered_map<observers::ArrayComponentId,
+                                 ExtentsAndTensorVolumeData>>*>
```
Maybe `db::item_type<...>` to make this shorter?

wthrowe commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -0,0 +1,163 @@
+          auto& volume_file = h5file.try_insert<h5::VolumeData>("/element_data",
+                                                                version_number);
+          for (const auto& id_and_tensor_data_for_grid :
+               volume_data->at(observation_id)) {
```
Race condition: This can run concurrently with inserting a new element into the map in the previous action.

wthrowe commented on tests/Unit/IO/Observers/Test_VolumeObserver.cpp:
```diff
@@ -0,0 +1,202 @@
+  const auto grids = volume_file.list_grids(temporal_id);
+  const auto to_string_helper = [](const ElementId<2>& t) {
+    return MakeString{} << t;
+  };
```
`get_output<ElementId<2>>`

kidder replied:
check why a lambda?

wthrowe commented on tests/Unit/IO/Observers/Test_VolumeObserver.cpp:
```diff
@@ -0,0 +1,202 @@
+                        }));
+    const observers::ArrayComponentId array_id(
+        std::add_pointer_t<element_comp>{nullptr},
+        Parallel::ArrayIndex<std::decay_t<ElementIndex<2>>>{
```
Why the `decay_t`?

wthrowe commented on src/Evolution/Executables/ScalarWave/CMakeLists.txt:
```diff
@@ -6,6 +6,7 @@ set(LIBS_TO_LINK
     DiscontinuousGalerkin
     DomainCreators
     Informer
+    IO
```
'O' < 'n'

wthrowe commented on src/Evolution/Systems/ScalarWave/Actions.hpp:
```diff
@@ -0,0 +1,117 @@
+    // We hard-code the writing to large time values to avoid breaking the
+    // tests.
+    if (time_by_timestep_value % 1000 == 0 and time_by_timestep_value > 0) {
+      // Get an Index of the dimensionality
```
I don't understand this comment.

nilsdeppe replied:
Yea, me either... Best guess is code changed and I forgot to remove/update comment....

wthrowe commented on src/Evolution/Systems/ScalarWave/Actions.hpp:
```diff
@@ -0,0 +1,117 @@
+
+      // Remove tensor types, only storing individual components.
+      std::vector<TensorComponent> components;
+      components.reserve(2 * Dim + 2);
```
I count `3 * Dim + 4`.

wthrowe commented on src/Evolution/Systems/ScalarWave/Actions.hpp:
```diff
@@ -0,0 +1,117 @@
+      const auto exact_solution = Parallel::get<solution_tag>(cache).variables(
+          inertial_coordinates, time.value(), typename Vars::tags_list{});
+      DataVector error =
+          tuples::get<ScalarWave::Psi>(exact_solution).get() - psi.get();
```
Move to immediately before use, since that pattern is followed for all the reuses.

wthrowe commented on src/Evolution/Systems/ScalarWave/Actions.hpp:
```diff
@@ -0,0 +1,117 @@
+ * - Observation frequency is currently hard-coded and must manually be updated.
+ *   Look for `time_by_timestep_value` to update.
+ * - Writes the solution \f$\Psi, \Pi\f$, and \f$\Phi_i\f$, and the error in
+ *   \f$\Psi\f$ and \f$\Pi\f$ to disk.
```
It does all three errors.

wthrowe reviewed (changes requested):
The tests Unit.IO.Observers.VolumeObserver and Unit.IO.H5.VolumeData fail on my system, both with the error
```
############ ERROR ############
Node: 0 Proc: 0
Line: 196 of /home/wthrowe/spectre/src/IO/H5/Helpers.cpp
Function: std::vector<T> h5::read_rank1_attribute(const hid_t, const std::string &) [T = unsigned long]
The read HDF5 type of the attribute (50331742) has a different size than the type we are reading. The stored size is 8 while the expected size is 8
############ ERROR ############
```
with relevant stack frames
```
  [0:3] std::vector<unsigned long, std::allocator<unsigned long> > h5::read_rank1_attribute<unsigned long>(int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x3e2  [0x43f4db2]
  [0:4] h5::VolumeData::get_extents(unsigned long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0x14f  [0x43fe48f]
```

wthrowe commented:
For the test failures, it looks like the error check is a bug and just needs to be changed to match what is printed in the message.

nilsdeppe committed:
Add support for calling threaded actions remotely

Previously threaded actions could only be called on the local object, but we
want to be able to call them remotely. One case where this is useful is when
wanting to write data to disk on N nodes where N is less than the total number
of nodes being used. In this case we invoke a threaded action on node M which
starts writing to disk. At the same time nodes can still contribute their data
at different ObservationIds to the nodegroup element M using threaded
actions. Another use case is being able to queue a threaded action with the
runtime system so that we can relinquish control back to the runtime system and
let proper threading occur.

wthrowe commented:
One of the test failures already exists on develop, so I'm going to include a fix in a small PR.

nilsdeppe committed:
Add support for threaded_actions to mocking framework

nilsdeppe committed:
Rename InvokeSimpleActionBase to InvokeActionBase

nilsdeppe committed:
Add ObserverWriter component

Use a group for reducing data on each processing element to avoid having to use
locks, then use locks for reducing the data to the nodegroup from the group. The
nodegroup will then reduce further or write to disk. Currently we write one file
per node.

nilsdeppe commented:
Okay, I updated this. There were some fairly major changes:
- get threaded actions into the mocking framework
- get threaded actions working via remote calls
- use a group and a nodegroup to reduce the amount of time we spend dealing with locks.

kidder commented on cmake/SetupCharmAlgorithm.py:
```diff
@@ -45,6 +45,15 @@ def create_interface_file(args):
               "    entry void perform_algorithm();\n" \
               "\n" % (args['algorithm_name'], args['algorithm_name'])
 
+    if (args['algorithm_type'] == "nodegroup"):
```
explicitly state what is obvious in the commit message

kidder commented on src/Parallel/Algorithm.hpp:
```diff
@@ -214,7 +215,24 @@ class AlgorithmImpl<ParallelComponent, ChareType, Metavariables,
+  template <typename Action>
+  void threaded_action() noexcept {
+    // NOLINTNEXTLINE(modernize-redundant-void-arg)
+    (void)Parallel::charmxx::RegisterThreadedAction<ParallelComponent,
```
make issue to write our own NodeLock class

kidder commented on tests/Unit/ActionTesting.hpp:
```diff
@@ -567,6 +651,29 @@ class MockProxy {
+  void threaded_action(std::tuple<Args...> args) noexcept {
+    if (local_algorithms_->size() != 1) {
+      ERROR(
+          "NodeGroups must have exactly one element in the during testing, but "
```
drop "in the"

kidder commented on tests/Unit/ActionTesting.hpp:
```diff
@@ -567,6 +651,29 @@ class MockProxy {
+  void threaded_action() noexcept {
+    if (local_algorithms_->size() != 1) {
+      ERROR(
+          "NodeGroups must have exactly one element in the during testing, but "
```
ditto

kidder commented on tests/Unit/IO/Observers/ObserverHelpers.hpp:
```diff
@@ -0,0 +1,69 @@
+// clang-tidy: We want all the classes in this header file to have internal
+// linkage in whatever object file they're being compiled, therefore we use an
+// anonymous namespace in the header.
+namespace {  // NOLINT
```
name this

kidder commented on src/IO/Observer/VolumeActions.hpp:
```diff
@@ -0,0 +1,227 @@
+namespace ThreadedActions {
+/*!
+ * \ingroup ObserverGroup
+ * \brief Writes volume data at the `temporial_id` "time" to disk.
```
temporal_id

kidder commented:
@wthrowe this should be ready (after a squash)

nilsdeppe commented:
Okay, rebased and squashed.
